File: schemas/compliance-grc/drizzle/audits.ts
// audits: Compliance audits and assessments with type, status, scope, and scheduling.
// See README.md for full design rationale.

import { pgTable, uuid, text, timestamp, index, pgEnum } from "drizzle-orm/pg-core";
import { organizations } from "./organizations";
import { users } from "./users";

export const auditTypeEnum = pgEnum("audit_type", [
  "internal",
  "external",
  "self_assessment",
  "certification",
]);

export const auditStatusEnum = pgEnum("audit_status", [
  "planned",
  "in_progress",
  "review",
  "completed",
  "cancelled",
]);

export const audits = pgTable(
  "audits",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    organizationId: uuid("organization_id").references(() => organizations.id, { onDelete: "cascade" }),
    leadAuditorId: uuid("lead_auditor_id").references(() => users.id, { onDelete: "set null" }),
    title: text("title").notNull(),
    auditType: auditTypeEnum("audit_type").notNull(),
    status: auditStatusEnum("status").notNull().default("planned"),
    scope: text("scope"),
    startDate: text("start_date"),
    endDate: text("end_date"),
    conclusion: text("conclusion"),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    index("idx_audits_organization_id").on(table.organizationId),
    index("idx_audits_lead_auditor_id").on(table.leadAuditorId),
    index("idx_audits_audit_type").on(table.auditType),
    index("idx_audits_status").on(table.status),
  ]
);

File: schemas/compliance-grc/spacetimedb/audits.rs
// audits: Compliance audits with type, scope, and lifecycle tracking.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

#[derive(SpacetimeType, Clone)]
pub enum AuditType {
    Internal,       // type: String
    External,
    SelfAssessment,
    Certification,
}

#[derive(SpacetimeType, Clone)]
pub enum AuditStatus {
    Planned,    // type: String
    InProgress,
    Review,
    Completed,
    Cancelled,
}

#[spacetimedb::table(name = audits, public)]
pub struct Audit {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub organization_id: Option<String>, // UUID — FK → organizations.id (cascade delete)

    #[index(btree)]
    pub lead_auditor_id: Option<String>, // UUID — FK → users.id (set null)

    pub title: String,

    #[index(btree)]
    pub audit_type: AuditType,

    #[index(btree)]
    pub status: AuditStatus, // Defaults to Planned.

    pub scope: Option<String>,

    pub start_date: Option<String>,

    pub end_date: Option<String>,

    pub conclusion: Option<String>,

    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
